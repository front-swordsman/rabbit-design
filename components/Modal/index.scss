@import '../styles/themes/default';

$modal: 'snake-modal';
$confirm: 'snake-confirm';

.#{$modal} {
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  z-index: $zindexModal;
  outline: 0;
  display: flex;
  align-items: center;
  justify-content: center;

  &-mask {
    position: fixed;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    background: rgba(0, 0, 0, 0.45);
    z-index: $zindexModal;
  }

  &-wrapper {
    background: #fff;
    border-radius: $borderRadius;
    min-width: 500px;
    position: relative;
    color: rgba(0, 0, 0, 0.65);

    &-inner {
      padding: 24px;
      font-size: $body;
      line-height: $lineHeight;
      word-wrap: break-word;
    }
  }

  &-header {
    padding: 16px 24px;
    border-bottom: 1px solid #e8e8e8;
  }

  &-title {
    color: rgba(0, 0, 0, 0.85);
    font-weight: 500;
    font-size: 16px;
    line-height: 22px;
  }

  &-close {
    position: absolute;
    right: 0;
    top: 0;
    color: rgba(0, 0, 0, 0.45);
    width: 55px;
    height: 55px;
    line-height: 55px;
    text-align: center;
    cursor: pointer;

    .snake-icon {
      font-size: 16px;
    }
  }

  &-body {
    &-contain {
      padding: 24px;
      font-size: 14px;
      line-height: 1.5;
      word-wrap: break-word;
    }
  }

  &-footer {
    padding: 10px 16px;
    text-align: right;
    border-top: 1px solid #e8e8e8;

    .snake-button {
      &:last-child {
        margin-left: 8px;
      }
    }
  }

  &-fade-appear,
  &-fade-enter {
    opacity: 0;
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: linear;
    animation-play-state: paused;
  }

  &-fade-appear-active,
  &-fade-enter-active {
    animation-name: modalFadeIn;
    animation-play-state: running;
  }

  &-fade-exit {
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: linear;
    animation-play-state: paused;
  }

  &-fade-exit-active {
    animation-name: modalFadeOut;
    animation-play-state: running;
  }

  &-zoom-appear,
  &-zoom-enter {
    opacity: 0;
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: $easeOutCirc;
    animation-play-state: paused;
  }

  &-zoom-appear-active,
  &-zoom-enter-active {
    animation-name: modalZoomIn;
    animation-play-state: running;
  }

  &-zoom-exit {
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: $easeOutCirc;
    animation-play-state: paused;
  }

  &-zoom-exit-active {
    animation-name: modalZoomOut;
    animation-play-state: running;
  }
}

.#{$confirm} {
  .#{$modal}-wrapper {
    min-width: 400px;
  }

  .#{$modal}-wrapper-inner {
    padding: 32px 32px 24px;
  }

  &-body-header {
    display: flex;
    align-items: center;
  }

  &-title {
    margin-left: 16px;
    color: rgba(0, 0, 0, 0.85);
  }

  &-content {
    margin-left: 40px;
    margin-top: 10px;
    color: rgba(0, 0, 0, 0.65);
  }

  &-btn {
    text-align: right;

    .snake-button {
      &:last-child {
        margin-left: 8px;
      }
    }
  }
}

@keyframes modalFadeIn {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

@keyframes modalFadeOut {
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}

@keyframes modalZoomIn {
  0% {
    opacity: 0;
    transform: scale(0.2);
  }

  100% {
    opacity: 1;
    transform: scale(1);
  }
}

@keyframes modalZoomOut {
  0% {
    transform: scale(1);
  }

  100% {
    opacity: 0;
    transform: scale(0.2);
  }
}
