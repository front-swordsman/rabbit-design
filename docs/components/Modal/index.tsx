import * as React from 'react'
import ModalDoc from './README.md'

export default function ModalView() {
  return <ModalDoc />
}
