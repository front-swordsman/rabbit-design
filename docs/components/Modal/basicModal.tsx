import * as React from 'react'
import { Button, Modal } from 'components'

export default function BasicModal() {
  const [visible, setVisible] = React.useState(false)
  return (
    <div>
      <h1>基础 Modal</h1>
      <Button onClick={() => setVisible(true)}>基础用法</Button>
      <Modal
        visible={visible}
        onCancel={() => setVisible(false)}
        title="基础 Modal "
        onOk={() => setVisible(false)}
      >
        <div>基础modal内容区</div>
      </Modal>
    </div>
  )
}
