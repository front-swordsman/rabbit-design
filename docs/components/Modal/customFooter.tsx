import * as React from 'react'
import { Button, Modal } from 'components'

export default function BasicModal() {
  const [visible, setVisible] = React.useState(false)
  const [loading, setLoading] = React.useState(false)
  return (
    <div>
      <h1>自定义底部</h1>
      <Button onClick={() => setVisible(true)}>自定义底部</Button>
      <Modal
        visible={visible}
        onCancel={() => setVisible(false)}
        title="基础 Modal "
        onOk={() => setVisible(false)}
        footer={
          <>
            <Button type="gray" onClick={() => setVisible(false)}>
              返回
            </Button>
            <Button
              onClick={() => {
                setLoading(true)
                setTimeout(() => {
                  setLoading(false)
                  setVisible(false)
                }, 2000)
              }}
              loading={loading}
            >
              确认
            </Button>
          </>
        }
      >
        <div>基础modal</div>
      </Modal>
    </div>
  )
}
