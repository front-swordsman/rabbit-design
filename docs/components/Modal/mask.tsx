import * as React from 'react'
import { Button, Modal } from 'components'

export default function BasicModal() {
  const [visible, setVisible] = React.useState(false)
  return (
    <div>
      <h1>点击遮罩层不消失</h1>
      <Button onClick={() => setVisible(true)}>点击遮罩层不消失</Button>
      <Modal
        visible={visible}
        onCancel={() => setVisible(false)}
        title="基础 Modal "
        onOk={() => setVisible(false)}
        maskClosable={false}
      >
        <div>基础modal</div>
      </Modal>
    </div>
  )
}
