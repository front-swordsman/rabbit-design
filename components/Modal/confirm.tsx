import * as React from 'react'
import * as ReactDOM from 'react-dom'
import cx from 'classnames'
import Modal from './modal'
import Button from '../Button'
import Icon from '../Icon'
import { AlertProps } from 'types/modal'
import { noop } from '../utils/tool'

const prefixCls = 'snake-confirm'

export interface Close {
  close: () => void
}

export default function confirm({
  content,
  icon = <Icon type="question-circle" color="#faad14" />,
  okText = '确认',
  onOk = noop,
  okButtonProps,
  title,
  onCancel = noop,
  cancelButtonProps,
  cancelText = '取消',
  className,
  hasCancelBtn = true,
  ...rest
}: AlertProps): Close {
  const div = document.createElement('div')
  document.body.appendChild(div)

  let isClosed = false

  const getClassName = () => {
    return cx('snake-confirm', className)
  }

  const close = () => {
    ReactDOM.render(renderModal(false), div)
  }

  const destroy = () => {
    ReactDOM.unmountComponentAtNode(div)
    if (div && div.parentNode) {
      div.parentNode.removeChild(div)
    }
  }

  const handleClick = () => {
    // 可能会存在异步情况 finally 兼容性不是特别好
    if (isClosed) return
    Promise.resolve(onOk())
      .then(() => {
        isClosed = true
        close()
      })
      .catch(() => {
        isClosed = true
        close()
      })
  }

  const renderModal = (visible: boolean) => {
    return (
      <Modal
        {...rest}
        closable={false}
        onCancel={() => {
          onCancel()
          close()
        }}
        title={null}
        footer={null}
        visible={visible}
        afterClose={destroy}
        className={getClassName()}
      >
        <div className={`${prefixCls}-wrapper`}>
          <div className={`${prefixCls}-body`}>
            <div className={`${prefixCls}-body-header`}>
              {icon}
              {title ? <div className={`${prefixCls}-title`}>{title}</div> : null}
            </div>
            {content ? <div className={`${prefixCls}-content`}>{content}</div> : null}
          </div>
          <div className={`${prefixCls}-btn`}>
            {hasCancelBtn ? (
              <Button
                type="gray"
                onClick={() => {
                  onCancel()
                  close()
                }}
                {...cancelButtonProps}
              >
                {cancelText}
              </Button>
            ) : null}
            <Button onClick={handleClick} {...okButtonProps}>
              {okText}
            </Button>
          </div>
        </div>
      </Modal>
    )
  }

  ReactDOM.render(renderModal(true), div)

  return {
    close
  }
}

export function info({
  icon = <Icon type="info-circle" color="#1890ff" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: AlertProps) {
  return confirm({ icon, hasCancelBtn, okText, ...rest })
}

export function success({
  icon = <Icon type="check-circle" color="#52c41a" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: AlertProps) {
  return confirm({ icon, hasCancelBtn, okText, ...rest })
}

export function error({
  icon = <Icon type="close-circle" color="#f5222d" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: AlertProps) {
  return confirm({ icon, hasCancelBtn, okText, ...rest })
}

export function warning({
  icon = <Icon type="warning-circle" color="#faad14" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: AlertProps) {
  return confirm({ icon, hasCancelBtn, okText, ...rest })
}
