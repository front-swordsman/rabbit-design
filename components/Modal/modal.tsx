import * as React from 'react'
import cx from 'classnames'
import { ModalProps, AlertProps } from 'types/modal'
import Overlay from '../Overlay'
import Button from '../Button'
import confirm, { info, success, error, warning, Close } from './confirm'
import { noop } from '../utils/tool'

import './index.scss'

type Confirm = (props: AlertProps) => Close

const defaultProps = {
  cancelText: '取消',
  okText: '确定',
  closable: true,
  destroy: true,
  onOk: noop,
  onCancel: noop,
  visible: false,
  maskClosable: true,
  esc: true,
  center: false,
  afterClose: noop
}

const prefixCls = 'snake-modal'

const renderHeader = ({ title }: ModalProps) => {
  if (title) {
    return <div className={`${prefixCls}-title`}>{title}</div>
  }
  return null
}

const renderFooter = ({
  cancelText,
  okText,
  onOk,
  okButtonProps,
  cancelButtonProps,
  footer,
  onCancel
}: ModalProps) => {
  if (footer === null) return null
  return (
    <div>
      {footer || (
        <>
          <Button onClick={onCancel} {...cancelButtonProps} type="gray">
            {cancelText}
          </Button>
          <Button onClick={onOk} {...okButtonProps}>
            {okText}
          </Button>
        </>
      )}
    </div>
  )
}

const Modal: React.FC<ModalProps> & {
  confirm: Confirm
  info: Confirm
  success: Confirm
  error: Confirm
  warning: Confirm
} = modalProps => {
  const props = { ...defaultProps, ...modalProps }
  const {
    children,
    className,
    center,
    esc,
    destroy,
    onCancel,
    visible,
    style,
    maskClosable,
    closable,
    zIndex,
    width,
    afterClose
  } = props

  const getClassStr = React.useCallback(() => {
    return cx(
      {
        [`${prefixCls}-center`]: center
      },
      className
    )
  }, [className, center])

  const getStyle = React.useCallback(() => {
    const cloneStyle: React.CSSProperties = {}
    if (width) {
      cloneStyle.width = width
    }
    return { ...style, ...cloneStyle }
  }, [width, style])

  return (
    <Overlay
      prefixCls={prefixCls}
      footer={renderFooter(props)}
      header={renderHeader(props)}
      wrapperClassName={getClassStr()}
      esc={esc}
      destroy={destroy}
      onClose={onCancel}
      visible={visible}
      wrapperStyle={getStyle()}
      maskClosable={maskClosable}
      closable={closable}
      zIndex={zIndex}
      afterClose={afterClose}
    >
      {children}
    </Overlay>
  )
}

Modal.confirm = confirm
Modal.success = success
Modal.error = error
Modal.warning = warning
Modal.info = info

export default Modal
