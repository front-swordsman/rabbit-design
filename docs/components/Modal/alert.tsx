import * as React from 'react'
import Modal from 'components/Modal/index'
import { Button } from 'components'

const confirm = Modal.confirm
const success = Modal.success
const info = Modal.info
const error = Modal.error
const warning = Modal.warning

export default function Alert() {
  const closeRef = React.useRef<any>()
  return (
    <div>
      <h1>快捷调用</h1>
      <Button
        onClick={() => {
          closeRef.current = confirm({
            content: 'confirm内容区',
            title: 'Do you Want to delete these items?'
          })
          setTimeout(() => {
            closeRef.current.close()
          }, 4000)
        }}
      >
        confirm
      </Button>
      <Button
        onClick={() => info({ content: 'info内容区', title: 'Do you Want to delete these items?' })}
        style={{ marginLeft: 20 }}
      >
        info
      </Button>
      <Button
        onClick={() =>
          success({ content: 'success内容区', title: 'Do you Want to delete these items?' })
        }
        style={{ marginLeft: 20 }}
      >
        success
      </Button>
      <Button
        onClick={() =>
          error({ content: 'error内容区', title: 'Do you Want to delete these items?' })
        }
        style={{ marginLeft: 20 }}
      >
        error
      </Button>
      <Button
        onClick={() =>
          warning({ content: 'warning内容区', title: 'Do you Want to delete these items?' })
        }
        style={{ marginLeft: 20 }}
      >
        warning
      </Button>
    </div>
  )
}
