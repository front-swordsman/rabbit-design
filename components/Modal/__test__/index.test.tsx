import * as React from 'react'
import 'jest-dom/extend-expect'
import { render, fireEvent } from 'react-testing-library'
import Modal from '../index'
import BasicModal from '../../../docs/components/Modal/basicModal'
import Alert from '../../../docs/components/Modal/alert'

jest.mock('react-transition-group', () => {
  return {
    CSSTransition: props => {
      return <div>{props.in || !props.unmountOnExit ? props.children : null}</div>
    }
  }
})

describe('Modal Test', () => {
  it('modal visible snapshot', () => {
    const { baseElement } = render(<Modal visible>内容区</Modal>)
    expect(baseElement).toMatchSnapshot()
  })

  it('modal without footer', () => {
    const { baseElement } = render(
      <Modal visible footer={null}>
        内容区
      </Modal>
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('modal visible = false', () => {
    const { baseElement } = render(<Modal>内容区</Modal>)
    expect(baseElement).toMatchSnapshot()
  })

  it('click ok or cancel modal fade', () => {
    const { getByText, queryByText } = render(<BasicModal />)
    const btn = getByText(/基础用法/i) as HTMLButtonElement
    fireEvent.click(btn)
    expect(getByText(/基础modal内容区/i)).toBeInTheDocument()
    const okBtn = getByText(/确定/i) as HTMLButtonElement
    fireEvent.click(okBtn)
    expect(queryByText(/基础modal内容区/i)).not.toBeInTheDocument()
    fireEvent.click(btn)
    expect(getByText(/基础modal内容区/i)).toBeInTheDocument()
    const cancelBtn = getByText(/取消/i) as HTMLButtonElement
    fireEvent.click(cancelBtn)
    expect(queryByText(/基础modal内容区/i)).not.toBeInTheDocument()
  })

  it('custom btn', () => {
    const { getByText } = render(
      <Modal visible okText="自定义确认" cancelText="自定义取消">
        内容区
      </Modal>
    )
    expect(getByText(/自定义确认/i)).not.toBeNull()
    expect(getByText(/自定义取消/i)).not.toBeNull()
  })

  it('Modal confirm', () => {
    const { getByText, queryByText } = render(<Alert />)
    const btn = getByText(/confirm/i) as HTMLButtonElement
    fireEvent.click(btn)
    expect(getByText(/confirm内容区/i)).toBeInTheDocument()
    const okBtn = getByText(/确认/i) as HTMLButtonElement
    fireEvent.click(okBtn)
    process.nextTick(() => {
      expect(queryByText(/confirm内容区/i)).not.toBeInTheDocument()
    })
    fireEvent.click(btn)
    const cancelBtn = getByText(/取消/i) as HTMLButtonElement
    fireEvent.click(cancelBtn)
    process.nextTick(() => {
      expect(queryByText(/confirm内容区/i)).not.toBeInTheDocument()
    })
  })

  it('Modal info', () => {
    const { getByText } = render(<Alert />)
    const btn = getByText(/info/i) as HTMLButtonElement
    fireEvent.click(btn)
    expect(getByText(/info内容区/i)).toBeInTheDocument()
  })

  it('Modal success', () => {
    const { getByText } = render(<Alert />)
    const btn = getByText(/success/i) as HTMLButtonElement
    fireEvent.click(btn)
    expect(getByText(/success内容区/i)).toBeInTheDocument()
  })

  it('Modal warning', () => {
    const { getByText } = render(<Alert />)
    const btn = getByText(/warning/i) as HTMLButtonElement
    fireEvent.click(btn)
    expect(getByText(/warning内容区/i)).toBeInTheDocument()
  })

  it('Modal error', () => {
    const { getByText } = render(<Alert />)
    const btn = getByText(/error/i) as HTMLButtonElement
    fireEvent.click(btn)
    expect(getByText(/error内容区/i)).toBeInTheDocument()
  })
})
