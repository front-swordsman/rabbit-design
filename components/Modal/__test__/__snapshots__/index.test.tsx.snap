// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Test modal visible = false 1`] = `
<body>
  <script
    src="https://at.alicdn.com/t/font_1127944_82mztmm5t8t.js"
  />
  <div />
  <div />
  <div />
</body>
`;

exports[`Modal Test modal visible snapshot 1`] = `
<body>
  <div />
  <div>
    <div
      class="snake-modal-mask"
    />
  </div>
  <div>
    <div
      class="snake-modal"
      tabindex="-1"
    >
      <div
        class="snake-modal-wrapper"
      >
        <div
          class="snake-modal-close"
        >
          <svg
            class="snake-icon snake-icon-close"
          >
            <use
              xlink:href="#icon-close"
            />
          </svg>
        </div>
        <div
          class="snake-modal-wrapper-inner"
        >
          内容区
        </div>
        <div
          class="snake-modal-footer"
        >
          <div>
            <button
              class="snake-button snake-button-default snake-button-btn-gray"
              type="button"
            >
              <span>
                取消
              </span>
            </button>
            <button
              class="snake-button snake-button-default snake-button-btn-primary"
              type="button"
            >
              <span>
                确定
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <script
    src="https://at.alicdn.com/t/font_1127944_82mztmm5t8t.js"
  />
</body>
`;

exports[`Modal Test modal without footer 1`] = `
<body>
  <script
    src="https://at.alicdn.com/t/font_1127944_82mztmm5t8t.js"
  />
  <div />
  <div>
    <div
      class="snake-modal-mask"
    />
  </div>
  <div>
    <div
      class="snake-modal"
      tabindex="-1"
    >
      <div
        class="snake-modal-wrapper"
      >
        <div
          class="snake-modal-close"
        >
          <svg
            class="snake-icon snake-icon-close"
          >
            <use
              xlink:href="#icon-close"
            />
          </svg>
        </div>
        <div
          class="snake-modal-wrapper-inner"
        >
          内容区
        </div>
      </div>
    </div>
  </div>
</body>
`;
